Rework SDK and user `tsconfig.json` for optimal DX (#2656)
wasp-lang/wasp: waspc/ChangeLog.md

## Unreleased

### ⚠️ Breaking Changes

> Address all breaking changes by following [the official migration guide](https://wasp.sh/docs/migration-guides/migrate-from-0-16-to-0-17). This is just a short overview:

Reviewer: ```suggestion
Follow the [the official migration guide](https://wasp.sh/docs/migration-guides/migrate-from-0-16-to-0-17) to address all the breaking changes. Here's a short overview:
```

## Unreleased

### ⚠️ Breaking Changes
Address all breaking changes by following [the official migration guide](https://wasp.sh/docs/migration-guides/migrate-from-0-16-to-0-17). This is just a short overview:

- In the `usernameAndPassword` authentication method, the function `login()` imported from `wasp/client/auth` now accepts an object with `username` and `password` instead of two separate arguments ([#2598](https://github.com/wasp-lang/wasp/pull/2598))
- We've made some improvements to our TypeScript setup that require you to update the `tsconfig.json` file. The migration guide will lead you through them.

>   For posterity, here are all the non-cosmetic changes we've made:

Author: Haha, fair enough!

  ```json
  {
    "compilerOptions": {
      // ...
      "moduleDetection": "force",
      "isolatedModules": true,
      // Remove 'typeRoots'
    },
  }

>   ```

Reviewer 2: I'd rather express this as a `diff`, it renders nicely:

~~~suggestion
  ```diff
   {
     "compilerOptions": {
  +    "moduleDetection": "force",
  +    "isolatedModules": true,
  -    "typeRoots": [ ... ]
     },
   }
  ```
~~~

> - [The issue](https://github.com/testing-library/jest-dom/issues/546#issuecomment-1889884843) with `jest-dom` types is now addressed in `src/vite-env.d.ts`.

Reviewer 2: This entry is not very understandable for users I think, especially since it links to an issue in another project altogether, and the error it describes is not one our users were having (because we were already solving it). 

I'd go for something along the lines of

```suggestion
- The types for DOM helpers in tests are now imported from `src/vite-env.d.ts` instead of changing the `tsconfig.json`'s `typeRoots`.
```

(and I'm sure it can be improved further 😅)

### 🔧 Small improvements

- Show a friendlier error when there are no routes defined in the wasp file ([#2643](https://github.com/wasp-lang/wasp/pull/2643))

> - Better TypeScript support ([#2656](https://github.com/wasp-lang/wasp/pull/2656))

Reviewer 2: Let's be specific

```suggestion
- Modernized our TypeScript configuration and added support for jump-to-definition ([#2656](https://github.com/wasp-lang/wasp/pull/2656))
```
